### PR TITLE
CORS-3021: infra: aws: add wavelength zone support

### PR DESCRIPTION
**Pull request by r4f4**
Implement Wavelength zone support as it was done for the terraform provisioning in https://github.com/openshift/installer/pull/7369.

### CODE REVIEW

**Comment by r4f4** [comment]
/hold
Until https://github.com/openshift/release/pull/46270 is merged and we can run tests

**Comment by r4f4** [comment]
`altinfra-e2e-aws-ovn-wavelengthzones`:
```
time="2023-11-30T20:05:48Z" level=info msg="Created Carrier Gateway" carrier gateway=ci-op-lb8pqm07-e9ff3-bz5fm-cagw id=cagw-06573a7efc3d45cc4
time="2023-11-30T20:05:49Z" level=info msg="Created route table" carrier route table=ci-op-lb8pqm07-e9ff3-bz5fm-public-carrier id=rtb-0b8d2b601e93a2569
time="2023-11-30T20:05:49Z" level=info msg="Created route to Carrier gateway" carrier gateway=cagw-06573a7efc3d45cc4 carrier route table=ci-op-lb8pqm07-e9ff3-bz5fm-public-carrier id=rtb-0b8d2b601e93a2569
time="2023-11-30T20:05:50Z" level=info msg="Created subnet" subnet=ci-op-lb8pqm07-e9ff3-bz5fm-public-us-west-2-wl1-lax-wlz-1
time="2023-11-30T20:05:50Z" level=info msg="Created subnet" subnet=ci-op-lb8pqm07-e9ff3-bz5fm-private-us-west-2-wl1-lax-wlz-1
time="2023-11-30T20:05:51Z" level=info msg="Associated subnet with route table" subnet id=subnet-074b7921347a46e03 table id=rtb-08c47cbf8763db65f
time="2023-11-30T20:05:51Z" level=info msg="Associated subnet with route table" subnet id=subnet-0c99352150f0ff647 table id=rtb-0b8d2b601e93a2569
```
* Carrier gateway created
* Carrier route table to gateway created
* Public edge subnet associated with carrier route table because of wavelength zone.

**Comment by r4f4** [comment]
`altinfra-e2e-aws-ovn-localzones`:
```
time="2023-11-30T18:55:41Z" level=info msg="Created subnet" subnet=ci-op-lb8pqm07-c595f-jwcxg-public-us-east-1-bue-1a
time="2023-11-30T18:55:42Z" level=info msg="Created subnet" subnet=ci-op-lb8pqm07-c595f-jwcxg-private-us-east-1-bue-1a
time="2023-11-30T18:55:42Z" level=info msg="Associated subnet with route table" subnet id=subnet-0e78b8dbcfe68e788 table id=rtb-06dbc7a2614e3db35
time="2023-11-30T18:55:42Z" level=info msg="Associated subnet with route table" subnet id=subnet-0bd2639ec83f0fee9 table id=rtb-0bee6c27478ebe216
```
* No carrier gateway/route table created
* Public edge subnet associated with public route table.

**Comment by r4f4** [comment]
Update: `s/WavelengtyZoneType/WavelengthZoneType/g` since the type was renamed.

File: pkg/infrastructure/aws/vpc.go [hunk 1]
```diff
@@ -181,13 +202,31 @@ func createVPCResources(ctx context.Context, logger logrus.FieldLogger, ec2Clien
...
+
+			// Associate public subnet with either the public or carrier route
+			// table depending on the zone type
+			table = publicRouteTable
+			tableType := "public"
+			if zoneType, ok := vpcInput.edgeZonesType[zone]; ok && zoneType == awstypes.WavelengthZoneType {
+				table = carrierTable
+				tableType = "carrier"
+			}
+			publicSubnetID := publicZoneMap[zone]
```
Comment (mtulio) [inline pkg/infrastructure/aws/vpc.go:218]: Are we handling private installations in the altinfra version / this flow? I am also have doubts how the private/disconnected installations are handled in TF version as long we are also [checking only the zone length](https://github.com/openshift/installer/blob/1709379a76141e94113382f5514be6a4b5718e4b/data/data/aws/cluster/vpc/vpc-public.tf#L50-L51)

**Review by mtulio — COMMENTED** [review]
Overall looks good to me. I have a question related to the private installation flow.

File: pkg/infrastructure/aws/vpc.go (hunk 1 above)
Comment (r4f4) [reply pkg/infrastructure/aws/vpc.go:218]: I don't know how BYO VPCs related to the edge/wavelength zones though. Depending on what's expected, we might be missing a few bits.

Comment (mtulio) [reply pkg/infrastructure/aws/vpc.go:218]: @r4f4 , you are correct. I revisit the restricted[1]/private[2] and all uses BYO VPC path which will not touch that part. BYO VPC for WL follows the same pattern.
[1] https://docs.openshift.com/container-platform/4.14/installing/installing_aws/installing-restricted-networks-aws-installer-provisioned.html
[2] https://docs.openshift.com/container-platform/4.14/installing/installing_aws/installing-aws-private.html

**Comment by r4f4** [comment]
/hold
Will probably be obsoleted by the CAPI work.

**Comment by mtulio** [comment]
> @mtulio I think we are safe to abandon this given it was implemented in capi/capa. WDYT?

LGTM to close.

**Pull request closed without merging**